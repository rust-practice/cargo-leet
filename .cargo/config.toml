[alias]
i = "install --path . --features=tool"
lg = "run --features=tool -- leet gen"
