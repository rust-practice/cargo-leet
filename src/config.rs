pub struct Config {}

impl Config {
    pub const LEETCODE_URL: &str = "https://leetcode.com/problems/";
    pub const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}
