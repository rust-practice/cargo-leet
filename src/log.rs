use anyhow::Context;
use log::LevelFilter;
use log4rs::{
    append::console::ConsoleAppender,
    config::{Appender, Root},
    encode::pattern::PatternEncoder,
};

pub fn init_logging(log_level: LevelFilter) -> anyhow::Result<()> {
    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{h({M} {l} - {m})}{n}")))
        .build();

    let config = log4rs::Config::builder()
        .appender(Appender::builder().build("stdout", Box::new(stdout)))
        .build(Root::builder().appender("stdout").build(log_level))?;
    log4rs::init_config(config).context("Failed to initialize logging")?;
    Ok(())
}
