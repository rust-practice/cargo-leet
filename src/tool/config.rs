pub(crate) struct Config {}

impl Config {
    // assumed in the code using them URLs Must end with trailing "/"
    pub(crate) const LEETCODE_PROBLEM_URL: &'static str = "https://leetcode.com/problems/";
    pub(crate) const LEETCODE_GRAPH_QL: &'static str = "https://leetcode.com/graphql/";
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn problem_url_ends_with_slash() {
        // TODO On√®: Switch to using https://docs.rs/static_assertions/latest/static_assertions/
        assert!(Config::LEETCODE_PROBLEM_URL.ends_with('/'));
    }

    #[test]
    fn graph_ql_url_ends_with_slash() {
        assert!(Config::LEETCODE_GRAPH_QL.ends_with('/'));
    }
}
