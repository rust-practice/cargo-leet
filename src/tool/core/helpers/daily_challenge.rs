use anyhow::Context;
use log::info;

use crate::tool::config::Config;

use super::{get_response, local_store::path_local_store_daily_challenge};

#[derive(serde::Deserialize, Debug)]
struct DailyChallengeResponse {
    data: Data,
}
impl DailyChallengeResponse {
    fn into_title_slug(self) -> String {
        self.data
            .active_daily_coding_challenge_question
            .question
            .title_slug
    }
}

#[derive(serde::Deserialize, Debug)]
struct Data {
    #[serde(rename = "activeDailyCodingChallengeQuestion")]
    active_daily_coding_challenge_question: ActiveDailyCodingChallengeQuestion,
}

#[derive(serde::Deserialize, Debug)]
struct ActiveDailyCodingChallengeQuestion {
    question: Question,
}

#[derive(serde::Deserialize, Debug)]
struct Question {
    #[serde(rename = "titleSlug")]
    title_slug: String,
}

pub(crate) fn get_daily_challenge_slug() -> anyhow::Result<String> {
    info!("Attempting to get daily challenge");
    Ok(get_daily_challenge_response()?.into_title_slug())
}

fn get_daily_challenge_response() -> anyhow::Result<DailyChallengeResponse> {
    get_response(
        "unused_just_to_match_sig",
        local_store_request_daily_challenge,
        external_request_daily_challenge,
    )
}

fn local_store_request_daily_challenge(_needed_to_match_signature: &str) -> anyhow::Result<String> {
    let path = path_local_store_daily_challenge();
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_daily_challenge(_needed_to_match_signature: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request of daily challenge");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": "query questionOfToday {
                activeDailyCodingChallengeQuestion {
                    question {
                        titleSlug
                    }
                }
            }",
            "variables":{},
            "operationName":"questionOfToday"
        }))
        .context("get request for daily challenge failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests;
