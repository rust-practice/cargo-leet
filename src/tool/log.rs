#![cfg(feature = "tool")]
use env_logger::Builder;
use log::LevelFilter;

pub fn init_logging(level: LevelFilter) -> anyhow::Result<()> {
    Builder::new().filter(None, level).try_init()?;
    Ok(())
}
