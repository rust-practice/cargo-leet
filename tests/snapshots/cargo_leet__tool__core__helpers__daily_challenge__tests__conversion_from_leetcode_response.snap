---
source: src/tool/core/helpers/daily_challenge.rs
expression: get_daily_challenge_response().unwrap()
---
DailyChallengeResponse {
    data: Data {
        active_daily_coding_challenge_question: ActiveDailyCodingChallengeQuestion {
            question: Question {
                title_slug: "partition-array-for-maximum-sum",
            },
        },
    },
}
